[cp] override compute_log_sumexp to True for aten._scaled_dot_product_efficient_attention.default if False

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #145421

## Description
Our current CP doesn't support efficient attention when `compute_log_sumexp=False`. `compute_log_sumexp=False` only if that `requires_grad=False` and since PP's [shape inference](https://github.com/pytorch/pytorch/blob/d95a6babcc581ff06d1b914ee9f92c81b2e850e2/torch/distributed/pipelining/stage.py#L1387) happens under `torch.no_grad()` context , we need to override `compute_log_sumexp` to `True` in our CP attention implementation.

## Test
- Test PP+FSDP+CP w/ `mixed_precision = "float32"` in torchtitan

- `pytest test/distributed/tensor/test_attention.py -s -k test_ring_attention_sdpa`

Before:
<img width="1880" alt="image" src="https://github.com/user-attachments/assets/872ff583-295e-4751-a280-cf7f2d41c61a" />

After:
<img width="2988" alt="image" src="https://github.com/user-attachments/assets/4bdcc2e5-22a5-427a-91a5-82206d5bd78f" />


cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

XilunWu committed:
[pipeline] _shape_inferrence sets compute_log_sumexp=False for aten._scaled_dot_product_efficient_attention.default

[ghstack-poisoned]

XilunWu committed:
Update on "[pipeline] _shape_inferrence sets compute_log_sumexp=False for aten._scaled_dot_product_efficient_attention.default"


# Description
this `torch.no_grad()` sets `compute_log_sumexp=False` for the `aten._scaled_dot_product_efficient_attention.default` call and this is the case in which our CP code throws an exception. Would like to discuss if this `no_grad` is necessary.

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fegin commented:
Can we add a test with torch.no_grad?

H-Huang commented on torch/distributed/tensor/experimental/_attention.py:
```diff
@@ -225,8 +226,15 @@ def _scaled_dot_product_ring_efficient_attention(
+            f"be True but received {compute_log_sumexp}. "
+            "Override compute_log_sumexp to True.",
+        )
+        compute_log_sumexp = True
```
Any reason that it is a parameter to the function if it is always going to be `True`? 

Sorry, I don't have a lot of context on how this method is used, do you have pointers?

XilunWu replied:
This function is used in dispatcher when the sdp efficient attention ATEN op is called. So the function signature better stays consistent with the op (i.e. keep the compute_log_sumexp argument even if we require it be True):  https://github.com/pytorch/pytorch/blob/2f317bbdbc416086c907941593b97c947bbbf5d9/torch/distributed/tensor/experimental/_attention.py#L545-L549

Besides, I haven't looked into the reason why `compute_log_sumexp=False` gives error and hopefully we can make it compatible. But the bottom line is, setting `compute_log_sumexp=True` shouldn't have significant overhead. A possible exception is in inference. I'll leave a TODO here and mark this PR as a temporary fix to unblock torchtitan 1.0 release.

drisspg replied:
TBH I dont even think you need to warn here. There is no real way for users to control this and its an optimization to save some memory + some writes for the kernel when backwards wont be used. 

Neither of these are true for the CP use case since you will always need the LSE to merge. So I think just set it to true and dont worry

XilunWu committed:
Update on "[cp] override compute_log_sumexp to True for aten._scaled_dot_product_efficient_attention.default if False"


## Description
Our current CP doesn't support efficient attention when `compute_log_sumexp=False`. `compute_log_sumexp=False` only if that `requires_grad=False` and since PP's [shape inference](https://github.com/pytorch/pytorch/blob/d95a6babcc581ff06d1b914ee9f92c81b2e850e2/torch/distributed/pipelining/stage.py#L1387) happens under `torch.no_grad()` context , we need to override `compute_log_sumexp` to `True` in our CP attention implementation.

## Test
- Test PP+FSDP+CP w/ `mixed_precision = "float32"` in torchtitan

- `pytest test/distributed/tensor/test_attention.py -s -k test_ring_attention_sdpa`

Before:
<img width="1880" alt="image" src="https://github.com/user-attachments/assets/872ff583-295e-4751-a280-cf7f2d41c61a" />

After:
<img width="2988" alt="image" src="https://github.com/user-attachments/assets/4bdcc2e5-22a5-427a-91a5-82206d5bd78f" />


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

tianyu-l reviewed (commented):
The PR itself looks ok to me. But I wonder what happens if efficient attention is used with `torch.no_grad` but without CP. Wouldn't `compute_log_sumexp` be set to `False` and cause similar issues? cc: @drisspg

fegin commented:
@tianyu-l No, it won't fail. The reason we are seeing the failure is because CP requires log_sumexp to merge the results. This is orthogonal to whether the original SDPA needs log_sumexp or not.

H-Huang reviewed (approved):
Agree with @drisspg to remove the warn. Otherwise LGTM!

XilunWu committed:
Update on "[cp] override compute_log_sumexp to True for aten._scaled_dot_product_efficient_attention.default if False"


## Description
Our current CP doesn't support efficient attention when `compute_log_sumexp=False`. `compute_log_sumexp=False` only if that `requires_grad=False` and since PP's [shape inference](https://github.com/pytorch/pytorch/blob/d95a6babcc581ff06d1b914ee9f92c81b2e850e2/torch/distributed/pipelining/stage.py#L1387) happens under `torch.no_grad()` context , we need to override `compute_log_sumexp` to `True` in our CP attention implementation.

## Test
- Test PP+FSDP+CP w/ `mixed_precision = "float32"` in torchtitan

- `pytest test/distributed/tensor/test_attention.py -s -k test_ring_attention_sdpa`

Before:
<img width="1880" alt="image" src="https://github.com/user-attachments/assets/872ff583-295e-4751-a280-cf7f2d41c61a" />

After:
<img width="2988" alt="image" src="https://github.com/user-attachments/assets/4bdcc2e5-22a5-427a-91a5-82206d5bd78f" />


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

XilunWu commented:
@pytorchbot merge
